src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,795 @@
+  template<typename xpu, typename DType>
+  static inline void Fill(mshadow::Stream<xpu> *s,
+                          const DType val,
+                          const OpReqType req,
+                          const TBlob& blob) {
+    using namespace mxnet_op;
+    using namespace mshadow::expr;
+    MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+      Kernel<MapSetToScalar<Req>, xpu>::Launch(s, blob.Size(), blob.dptr<DType>(), val);
+    });
   }
-  if (std::is_same<ROP, mshadow_op::identity>::value && req[1] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, rgrad_dptr);
-  } else if (req[1] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[1], Req,
-      {Kernel<BinaryOpBackwardUseNone<ROP, Req>, xpu>::Launch(s, size, rgrad_dptr,
-        ograd_dptr);});
+
+  /*! \brief Fill contiguous dense output rows with value computed from 0 lhs and 0 rhs input */
+  template<typename xpu, typename DType, typename OP>
+  static inline size_t FillDense(mshadow::Stream<xpu> *s,
+                                 const size_t idx_l,
+                                 const size_t idx_r,
+                                 const OpReqType req,
+                                 mshadow::Tensor<xpu, 2, DType> *out,
+                                 const size_t iter_out) {
+    using namespace mxnet_op;
+    using namespace mshadow::expr;
+    const int index_out_min = std::min(idx_l, idx_r);
+    if (index_out_min > iter_out) {
+      const size_t size = (*out)[iter_out].shape_.Size();
+      const DType zero_input_val = OP::Map(DType(0), DType(0));
+      #pragma omp parallel for
+      for (int i = static_cast<int>(iter_out); i < index_out_min; ++i) {
+        MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+          Kernel<MapSetToScalar<Req>, xpu>::Launch(s, size, (*out)[i].dptr_, zero_input_val);
+        });
+      }
+    }
+    return static_cast<size_t>(index_out_min);
   }
-}
-
-// TODO(haibin) This is a single-thread inefficient implementation
-// This implementation only works on CPU
-template<typename xpu, typename OP>
-void BinaryComputeRspRspImpl(const nnvm::NodeAttrs& attrs,
-                             const OpContext& ctx,
-                             const std::vector<NDArray>& inputs,
-                             const std::vector<OpReqType>& req,
-                             const std::vector<NDArray>& outputs) {
-  if (req[0] == kNullOp) return;
-  CHECK(req[0] == kWriteTo) << "only kWriteTo is supported for rowsparse elemwise_add";
-  using namespace rowsparse;
-  using namespace mshadow;
-  auto &lhs = inputs[0];
-  auto &rhs = inputs[1];
-  auto &output = outputs[0];
-
-  bool init_l = lhs.storage_initialized();
-  bool init_r = rhs.storage_initialized();
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  // both inputs are zeros
-  if (!init_l && !init_r) {
-    NDArray out = output;
-    FillZerosRspImpl(s, &out);
-    return;
+
+  template<typename DType>
+  static inline bool IsSameArray(const NDArray *a1, const NDArray *a2) {
```

Comment:
Rule of thumb is always passing const ref if possible for non-primitive objects, and passing pointer if have too.
1. This eliminates the need of checking nullptr.
2. No dilemma for defining the result when a1=a2=nullptr.